DS-4166 Index workspace, workflow and tasks in SOLR

abollini opened the pull request:
This PR is to share the implementation of the indexing features required for the implementation of the MyDSpace.

Integration Tests have been added to the DiscoverControllerIT to demonstrate:
1) inprogress items and tasks not popup elsewhere
2) the special **workspace** configuration works as expected (the user only see her own submissions)
3) the special **workflow** configuration works as expected (reviewers see only tasks that they have claimed or can claim)

 PR to highlight the small changes required to the Discover (search) endpoint https://github.com/DSpace/Rest7Contract/pull/55 
Namely the change of the embedded dspaceObject to rObject (result object) as now also workspace, workflow and tasks are returned

TO DO:
- move this PR to the DSpace project after than the DS-3851 will be merged

benbosman commented:
@abollini I was planning to review this PR, but it's currently unclear to me what this PR is supposed to do.
I can see mentions of MyDSpace and mentions of Solr indexing.
Is there a RestContract which explains the main goal of this PR, or can you add a more detailed description.
I also didn't see any ITs (at least not from the file filter), so that didn't help me either on identifying the functionality being implemented in this PR.

abollini commented:
@benbosman as anticipated in the last meeting this PR is now ready to be reviewed. It is in the 4Science repository just to reduce the amount of code to review as it is based on the workflow endpoint PR.

In this PR we are exposing inprogress submissions (workspace and workflow items) and tasks (pool and claimed) over the existing discover endpoint. The mydspace on the angular side just make a standard search using a special configuration "workspace" or "workflow".

abollini commented:
updated to work with the latest changes in https://github.com/DSpace/DSpace/pull/2312

benbosman commented:
After reading @KevinVdV's review https://github.com/DSpace/DSpace/pull/2312#pullrequestreview-211087415 and starting to review this PR, I noticed a weird starting point for indexing the items in the workspace and workflow

It seems additional types are created in https://github.com/4Science/DSpace/blob/DS-3851_workflow_new/dspace-api/src/main/java/org/dspace/core/Constants.java#L61 to filter in discovery using https://github.com/4Science/DSpace/blob/7f943837191265958eb724fac638b31066f239bf/dspace/config/spring/api/discovery.xml#L345

It also seems Angular will need to differentiate between an archived item, workspace item, workflow item, claimed task, pooled task based on the self link in the results.
I can see that a workflow item is returned when requesting the workspace configuration, and a claimed task or pooled task is returned when requesting the workflow configuration. This is also a very strange setup since both are in fact the same item.

It would make much more sense to me to index the Item object instead of the workspace item, workflow item, claimed task, pooled task, especially because workflow item overlaps with claimed task, pooled task. 
If the Item would be in the workflow, additional properties can be indexed and returned.
These additional properties can be used for the workspace and workflow discovery configuration.

This would avoid creating the additional types, the confusion of `BrowsableObject` pointed out by Tim and Kevin, the additional confusion because workflow item overlaps with claimed task, pooled task, …

abollini commented:
Hi @benbosman 
thanks to find the time to start this review.
**Workflowitem** are included in the **workspace** configuration because in your workspace you expect to see what you have submitter regardless to what is the publication status (workspace, workflow, published)

The **workflow configuration** (we can change the name to make this more obvious) allows the reviewers to focus on their tasks, so it is perfectly coherent to return pool and claimed tasks. As the configurable workflow allows local extensions we cannot never exclude that for a single workflowitem two tasks exist also for the same reviewer. Also ignoring this cutting edge scenario working directly with what is the focus of the user action allow to configure facets specific for the tasks (such as the required actiontype) that could be not possible if the information will be flat in the item as multiple parallel review of the same item is an obvious use case (also supported out-of-box by the configurable workflow)

abollini commented:
Updated with the latest changes in https://github.com/DSpace/DSpace/pull/2312 and renamed the rObject attribute in resultObject as by discussion in https://github.com/DSpace/Rest7Contract/pull/55

benbosman commented:
I'm summarizing the conceptual changes we agreed on for this PR:
* We've agreed to rename BrowsableObject to IndexableObject in Java
* We've agreed to rename "resultObject" to "indexableObject" in the discover REST result

The changes to the Constants class are currently under discussion:
* My proposal is not to change anything for the DSpaceObject Constants to avoid any big conceptual changes. So an item can still be found using `search.resourcetype:2`
* My proposal is limited to:
  * ensuring **new IndexableObject objects** such as the PoolTask would not use an integer, but rather be found using e.g. `search.objecttype:PoolTask`
  * an item can **also** be found using `search.objecttype:Item` but `search.resourcetype:2` is also supported
  * This would imply the `search.resourcetype` becomes deprecated but is not removed yet. Removing it would be part of DSpace 8.
  * This is a small change, and avoids the Constants to contain a list of all implementations of the IndexableObject with an assigned Integer
  * This ensures the new `IndexableObject` model which was introduced in https://github.com/DSpace/DSpace/pull/2312 can hereafter be used to index e.g. Metadata Schemas and Fields, a use case which would allow us to search for all fields containing "author" in the schema, element, qualifier or note for the admin edit item.

We've agreed that refactoring the `SolrServiceImpl` is not in scope of this PR because it's not a small change:
* The SolrServiceImpl now indexes a `PoolTask` using `indexInProgressSubmissionItem(Context context, Item item)` while the Item itself is not indexed
* The SolrServiceImpl now has to implement an indexing method for each `IndexableObject`
* A separate ticket will be created for this modification

Now that the scope of this PR is determined, I will also start the detailed reviewing.

benbosman commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -1422,13 +1585,150 @@ public int compare(MetadataValue mdv1,MetadataValue mdv2) {
+                    SolrInputDocument claimDoc = doc.deepCopy();
+                    addBasicInfoToDocument(claimDoc, Constants.WORKFLOW_CLAIMED, claimedTask.getID(), null, locations);
+                    claimDoc.addField("workflow.action", claimedTask.getActionID());
+                    claimDoc.addField("workflow.step", claimedTask.getStepID());
```
Does this imply, when faceting on the step, the facets will contain the step IDs, and not human-readable step names? Or is a different field used for this purpose?

abollini replied:
I have reduced the number of fields used as many were redundant. in these field we will have the action and step ID that are "string" so readable but not really ready to be displayed to the user. They are not really needed, nevertheless is quite convenient to have them immediately available in the solr document if you need to perform some analysis / check or manual report

KevinVdV commented on dspace/config/spring/api/discovery.xml:
```diff
@@ -289,6 +301,219 @@
+                </property>
+            </bean>
+        </property>
+        <property name="moreLikeThisConfiguration">
```
Can this property not be left out, I don't think we need the more like this configuration for workspace items.

abollini replied:
removed, see https://github.com/4Science/DSpace/pull/65/commits/cc11f4955677eaedfae70c54eb08e940bb6067fe

KevinVdV commented on dspace/config/spring/api/discovery.xml:
```diff
@@ -289,6 +301,219 @@
+        </property>
+        <!--Default result per page  -->
+        <property name="defaultRpp" value="10" />
+        <property name="hitHighlightingConfiguration">
```
Can this property not be left out, I don't think we need the hit highlighting configuration for workspace items.

abollini replied:
removed, see https://github.com/4Science/DSpace/pull/65/commits/cc11f4955677eaedfae70c54eb08e940bb6067fe

KevinVdV commented on dspace/config/spring/api/discovery.xml:
```diff
@@ -289,6 +301,219 @@
+			</list>
+		</property>
+        <!--The configuration for the recent submissions-->
+        <property name="recentSubmissionConfiguration">
```
Can this property not be left out, I don't think we need the recent submission configuration for workspace items.

abollini replied:
removed, see https://github.com/4Science/DSpace/pull/65/commits/cc11f4955677eaedfae70c54eb08e940bb6067fe

KevinVdV commented on dspace/config/spring/api/discovery.xml:
```diff
@@ -289,6 +301,219 @@
+			</list>
+		</property>
+        <!--The configuration for the recent submissions-->
+        <property name="recentSubmissionConfiguration">
```
Can this property not be left out, I don't think we need the recent submissions configuration for workflow items.

abollini replied:
removed, see https://github.com/4Science/DSpace/pull/65/commits/cc11f4955677eaedfae70c54eb08e940bb6067fe

KevinVdV commented on dspace/config/spring/api/discovery.xml:
```diff
@@ -289,6 +301,219 @@
+        </property>
+        <!--Default result per page  -->
+        <property name="defaultRpp" value="10" />
+        <property name="hitHighlightingConfiguration">
```
Can this property not be left out, I don't think we need the hit highlighting configuration for workflow items.

abollini replied:
removed, see https://github.com/4Science/DSpace/pull/65/commits/cc11f4955677eaedfae70c54eb08e940bb6067fe

KevinVdV commented on dspace/config/spring/api/discovery.xml:
```diff
@@ -289,6 +301,219 @@
+                </property>
+            </bean>
+        </property>
+        <property name="moreLikeThisConfiguration">
```
Can this property not be left out, I don't think we need the more like this configuration for workflow items.

abollini replied:
removed, see https://github.com/4Science/DSpace/commit/cc11f4955677eaedfae70c54eb08e940bb6067fe

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfigurationService.java:
```diff
@@ -20,9 +21,18 @@
  */
 public class DiscoveryConfigurationService {
 
+    private Map<String, List<String>> extraConfigurationMapping = new HashMap<String, List<String>>();
```
This configuration isn't used, the getters & setters aren't called, can you remove this variable.

abollini replied:
removed, see https://github.com/4Science/DSpace/pull/65/commits/cc11f4955677eaedfae70c54eb08e940bb6067fe

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/DiscoverResult.java:
```diff
@@ -28,8 +28,11 @@
+    private List<BrowsableObject> dspaceObjects;
     private Map<String, List<FacetResult>> facetResults;
+    private Map<String, List<FacetResult>> facetQueryResults;
+    private Map<String, List<FacetResult>> facetFieldResults;
```
Variable isn't used, neither getters nor setters are called. Can you remove it ?

abollini replied:
removed, see https://github.com/4Science/DSpace/pull/65/commits/cc11f4955677eaedfae70c54eb08e940bb6067fe

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/DiscoverResult.java:
```diff
@@ -28,8 +28,11 @@
-    private List<DSpaceObject> dspaceObjects;
+    private List<BrowsableObject> dspaceObjects;
     private Map<String, List<FacetResult>> facetResults;
+    private Map<String, List<FacetResult>> facetQueryResults;
```
Variable isn't used, neither getters nor setters are called. Can you remove it ?

abollini replied:
removed, see https://github.com/4Science/DSpace/pull/65/commits/cc11f4955677eaedfae70c54eb08e940bb6067fe

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/DiscoverResult.java:
```diff
@@ -178,28 +219,35 @@ public void setSpellCheckQuery(String spellCheckQuery) {
-        private DSpaceObject dso;
+        private BrowsableObject dso;
         private Map<String, List<String>> highlightResults;
+        private Map<String, List<String[]>> highlightResultsWithAuthority;
```
Variable isn't used, neither getters nor setters are called. Can you remove it ?

abollini replied:
removed, see https://github.com/4Science/DSpace/pull/65/commits/cc11f4955677eaedfae70c54eb08e940bb6067fe

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/IndexClient.java:
```diff
@@ -150,10 +168,55 @@ public static void main(String[] args) throws SQLException, IOException, SearchS
+        } else if (line.hasOption("item_uuid")) {
+            String itemUUID = line.getOptionValue("item_uuid");
+            Item item = ContentServiceFactory.getInstance().getItemService().find(context, UUID.fromString(itemUUID));
+            indexer.indexContent(context, item, line.hasOption("f"));
```
Best to add a check to see that the item was found, if it wasn't found a NullPointer will follow.

abollini replied:
done, see https://github.com/4Science/DSpace/pull/65/commits/8de7a50f5be1d2f85e37f8659decb772b5780ee1
same behavior than the existing -i option

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/IndexClient.java:
```diff
@@ -150,10 +168,55 @@ public static void main(String[] args) throws SQLException, IOException, SearchS
+                if (id.startsWith(ConfigurationManager.getProperty("handle.prefix")) || id.startsWith("123456789/")) {
+                    BrowsableObject dso = (BrowsableObject) HandleServiceFactory.getInstance()
+                            .getHandleService().resolveToObject(context, id);
+                    indexer.indexContent(context, dso, line.hasOption("f"));
```
Best to add a check to see that the object was found, if it wasn't found a NullPointer will follow.

abollini replied:
option not needed, removed (it was already supported by the -i )

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/IndexClient.java:
```diff
@@ -150,10 +168,55 @@ public static void main(String[] args) throws SQLException, IOException, SearchS
+        } else if (line.hasOption('e')) {
+            try {
+                String filename = line.getOptionValue('e');
+                FileInputStream fstream = new FileInputStream(filename);
```
Use try-with-resources or close this object in a "finally" clause.

abollini replied:
option out of scope for this PR, removed

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/IndexClient.java:
```diff
@@ -150,10 +168,55 @@ public static void main(String[] args) throws SQLException, IOException, SearchS
+                String filename = line.getOptionValue('e');
+                FileInputStream fstream = new FileInputStream(filename);
+                // Get the object of DataInputStream
+                DataInputStream in = new DataInputStream(fstream);
```
Use try-with-resources or close this object in a "finally" clause.

abollini replied:
option removed, the code doesn't exist anymore

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/IndexClient.java:
```diff
@@ -150,10 +168,55 @@ public static void main(String[] args) throws SQLException, IOException, SearchS
+                FileInputStream fstream = new FileInputStream(filename);
+                // Get the object of DataInputStream
+                DataInputStream in = new DataInputStream(fstream);
+                BufferedReader br = new BufferedReader(new InputStreamReader(in));
```
Use try-with-resources or close this object in a "finally" clause (for both the InputStreamReader & the BufferedReader).

abollini replied:
option removed, the code doesn't exist anymore

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/IndexClient.java:
```diff
@@ -150,10 +168,55 @@ public static void main(String[] args) throws SQLException, IOException, SearchS
+
+                int type = -1;
+                if (line.hasOption('t')) {
+                    type = Integer.parseInt(line.getOptionValue("t"));
```
This option isn't documented above.

abollini replied:
option removed as it is out of scope for this PR

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/IndexClient.java:
```diff
@@ -111,6 +127,8 @@ public static void main(String[] args) throws SQLException, IOException, SearchS
         options.addOption(OptionBuilder.isRequired(false).withDescription(
             "optimize search core").create("o"));
 
+        options.addOption("e", "readfile", true, "Read the identifier from a file");
```
These identifiers need to be items, best to alter the documentation here to make this clear.

abollini replied:
option removed as it is out of scope of this PR

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/IndexClient.java:
```diff
@@ -150,10 +168,55 @@ public static void main(String[] args) throws SQLException, IOException, SearchS
+                }
+                indexer.updateIndex(context, ids, line.hasOption("f"), type);
+            } catch (Exception e) {
+                log.error("Error: " + e.getMessage());
```
The error log here should also log the stacktrace.

abollini replied:
option removed, the code doesn't exist anymore

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -425,23 +452,44 @@ public void updateIndex(Context context) {
-                indexContent(context, community, force);
+    public void updateIndex(Context context, List<UUID> ids, boolean force, int type) {
+        if (type != Constants.ITEM) {
+            throw new RuntimeException("Only ITEM is supported in this mode - type founded: " + type);
```
Would an IllegalArgumentException not make more sense here instead of a generic "RuntimeException" as this is what it really is. An illegal Argument was passed along.

abollini replied:
the code doesn't exist anymore

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -425,23 +452,44 @@ public void updateIndex(Context context) {
+                    }
+                    break;
+                default:
+                    throw new RuntimeException("No type known: " + type);
```
Would an IllegalArgumentException not make more sense here instead of a generic "RuntimeException" as this is what it really is. An illegal Argument was passed along.

abollini replied:
done, https://github.com/4Science/DSpace/pull/65/commits/5e3164bd06e2eaefc0a047087b1ab7cebdb1ee4d

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -453,6 +501,40 @@ public void updateIndex(Context context, boolean force) {
+                ids.add(wfi.getItem().getID());
+            }
+        }
+        List<UUID>[] arrayIDList = Util.splitList(ids, numThreads);
```
If this method is called upon multiple times from the process the number of threads that can be active at the same time are not limited to the "numThreads" variable. This could really lead to memory leaks where we have way too many threads running at the same time. It would really be best to use a spring TaskExecutor: https://docs.spring.io/spring/docs/4.3.x/spring-framework-reference/html/scheduling.html. This will ensure that the number of threads started by the discovery process is limited to the "numThreads" variable

abollini replied:
support for multi-threads has been withdrawn from this PR as suggested by @cwilper https://github.com/4Science/DSpace/pull/65#issuecomment-477177919 and will be eventually introduced in a separate PR/ticket

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -453,6 +501,40 @@ public void updateIndex(Context context, boolean force) {
+            threads.add(thread);
+        }
+        boolean finished = false;
+        while (!finished) {
```
USe the spring TaskExecutors to find a better way to handle this.

abollini replied:
support for multi-threads has been withdrawn from this PR as suggested by @cwilper https://github.com/4Science/DSpace/pull/65#issuecomment-477177919 and will be eventually introduced in a separate PR/ticket

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -2305,4 +2597,64 @@ public String calculateExtremeValue(Context context, String valueField,
+        document.addField("namedresourcetype_keyword", fvalue);
+    }
+
+    class IndexerThread extends Thread {
```
Should really implement the Runnable class instead of extending thread as this is the preferred way to work with threads in java.

https://stackoverflow.com/questions/541487/implements-runnable-vs-extends-thread-in-java

abollini replied:
I agree with the suggestion but support for multi-threads has been withdrawn from this PR as suggested by @cwilper https://github.com/4Science/DSpace/pull/65#issuecomment-477177919 and will be eventually introduced in a separate PR/ticket

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -463,9 +545,27 @@ public void updateIndex(Context context, boolean force) {
+    public void cleanIndex(boolean force) throws IOException, SQLException, SearchServiceException {
+        if (force) {
+            try {
+                getSolr().deleteByQuery(
```
Why not use "*:*" as the query ?

abollini replied:
it is safe in case the local installation has customized the search index in a way that we don't know so that it includes now things maybe managed and indexed by other classes/services.

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -1931,6 +2182,20 @@ protected DiscoverResult retrieveResult(Context context, DiscoverQuery query, Qu
         return result;
     }
 
+    public DiscoverResult.FacetResult getDiscoveryFacet(Context context, FacetField facetField,
```
Method isn't used.

abollini replied:
removed https://github.com/4Science/DSpace/pull/65/commits/cc11f4955677eaedfae70c54eb08e940bb6067fe

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -2305,4 +2597,64 @@ public String calculateExtremeValue(Context context, String valueField,
+                    System.out.println(head + ":" + (idx++) + " / " + size);
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
```
Can this be removed, as it is already logged below.

abollini replied:
the code doesn't exist anymore

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -2305,4 +2597,64 @@ public String calculateExtremeValue(Context context, String valueField,
+                        indexContent(context, item, force);
+                        context.uncacheEntity(item);
+                    } catch (Exception ex) {
+                        log.error("ERROR: identifier item:" + id + " identifier thread:" + head);
```
Could you also log the stacktrace here ?

KevinVdV commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -2305,4 +2597,64 @@ public String calculateExtremeValue(Context context, String valueField,
+                    } catch (Exception ex) {
+                        log.error("ERROR: identifier item:" + id + " identifier thread:" + head);
+                    }
+                    System.out.println(head + ":" + (idx++) + " / " + size);
```
Can this end up in the logs instead of in the system out.

abollini replied:
he code doesn't exist anymore

cwilper commented:
Regarding the multi-threaded indexing, does it makes sense to do that as a distinct JIRA/PR pair? It doesn't seem essential to the functionality being introduced here, and would seem worth dedicated consideration.

I agree that moving toward something like spring executors would probably result in more robust thread management. I also think some dedicated testing would be good...at least to show/report that there's a throughput improvement for a typical site for the chosen default number of threads.

abollini commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceResourceRestrictionPlugin.java:
```diff
@@ -82,39 +89,66 @@ public void additionalIndex(Context context, DSpaceObject dso, SolrInputDocument
-                //Retrieve all the groups the current user is a member of !
+                boolean isWorkspace = StringUtils.startsWith(discoveryQuery.getDiscoveryConfigurationName(),
+                        "workspace");
+                boolean isWorkflow = StringUtils.startsWith(discoveryQuery.getDiscoveryConfigurationName(), "workflow");
```
you are right. Moreover, this is something required by the SolrServiceImpl and not an optional behaviour that can be enabled or disabled as the ResourceRestrictionPlugin. I have restored the previous implementation of the plugin (with a small modification to skip the workspace/workflow special configuration) and introduced the support for the required security directly in the SolrServiceImpl using dedicated fields instead than the read field

abollini replied:
See https://github.com/4Science/DSpace/pull/65/commits/006b938c4a97a7c43a8ca6fce9a717c6afb0ad69

abollini commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -1422,13 +1585,150 @@ public int compare(MetadataValue mdv1,MetadataValue mdv2) {
+
+            doc.addField("lastModified", item.getLastModified());
+            if (workspaceItem.getSubmitter() != null) {
+                doc.addField("read", "ws" + workspaceItem.getSubmitter().getID());
```
done, see https://github.com/4Science/DSpace/pull/65/commits/006b938c4a97a7c43a8ca6fce9a717c6afb0ad69

abollini commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -237,46 +260,45 @@ public void indexContent(Context context, DSpaceObject dso)
                          */
-                        unIndexContent(context, item);
-                        log.info("Removed Item: " + handle + " from Index");
+                        indexInProgressSubmissionItem(context, (Item) dso);
```
this is left for a follow-up PR as agreed

abollini commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java:
```diff
@@ -2305,4 +2597,64 @@ public String calculateExtremeValue(Context context, String valueField,
         return null;
     }
+
+    private void addNamedResourceTypeIndex(SolrInputDocument document, String acvalue, String fvalue) {
```
done, see https://github.com/4Science/DSpace/pull/65/commits/272f21a8dd0807c930c11410c4b5a70b0ea28c90

abollini commented:
Hi all, thanks to all of you @benbosman @KevinVdV @cwilper @tdonohue for your review. I tried to put response with links in all the comment to speedup the validation process.
Please note that the PR is now open on the official DSpace repository so we should continue here our discussion https://github.com/DSpace/DSpace/pull/2391

@benbosman  about your latest general comment https://github.com/4Science/DSpace/pull/65#issuecomment-475167938

> We've agreed to rename BrowsableObject to IndexableObject in Java

done, see https://github.com/4Science/DSpace/pull/65/commits/4b85bf4c564fd170b517af861055ac1e36a9e73b and https://github.com/4Science/DSpace/pull/65/commits/30899b0e555fe6ba51bfa5dade47f251ae713155

> We've agreed to rename "resultObject" to "indexableObject" in the discover REST result

done, see https://github.com/4Science/DSpace/pull/65/commits/b98d8f4d21266866d25aa4c78e3099a73884acdf


> We've agreed to make sure IndexableObject is implemented in Item, Collection, Community; and not in DSpaceObject

done, see https://github.com/4Science/DSpace/pull/65/commits/689ac4e3bd32aa155b1d752ad449d9bbdd9f9d1e


> the changes to the Constants class are currently under discussion

I have preferred to avoid any changes on that. If we will end in replace the search.resoucetype field with a search.resourcename field we can drop the new constants at this time instead than introduce a double way to deal with this things now.
I don't see any issue in the Constants class to have ID for not DSpaceObject, it is not stated in any place that the scope of the Constants class was limited to DSpaceObject, moreover also in the authorization infrastructure of the REST API we need constants to identify the different kind of objects regardless to be them dspaceobject or not, see
https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java#L81

If a change to the current implementation of the Constants class is needed it can be done in my opinion after the preview release or later.
